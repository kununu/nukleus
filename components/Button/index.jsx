import React, {Component, PropTypes} from 'react';
import classNames from 'classnames';

import styles from './index.scss';

export default class Button extends Component {
  static propTypes = {
    customTheme: PropTypes.string,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    link: PropTypes.element,
    mobileFullWidth: PropTypes.bool,
    onClick: PropTypes.func,
    outline: PropTypes.bool,
    text: PropTypes.string,
    type: PropTypes.oneOf(['default', 'primary', 'secondary', 'info', 'danger', 'link', 'custom'])
  };

  static defaultProps = {
    text: '',
    type: 'primary'
  };

  onClick = e => {
    e.preventDefault();

    if (this.props.disabled) return;

    this.props.onClick();
  }

  render () {
    const {
      customTheme,
      disabled,
      fullWidth,
      link,
      mobileFullWidth,
      outline,
      text,
      type
    } = this.props;

    const classes = classNames({
      [styles.button]: true,
      [styles.fullWidth]: fullWidth,
      [styles.mobileFullWidth]: mobileFullWidth,
      [styles.outline]: outline,
      [type === 'custom' ? customTheme : styles[type]]: true
    });

    const props = {
      className: classes,
      disabled,
      onClick: this.props.onClick && this.onClick
    };

    if (link) {
      return (
        <div className={disabled ? styles.disabledLink : ''}>
          {React.cloneElement(link, props)}
        </div>
      );
    }

    return (
      <button
        {...props}>
        {text}
      </button>
    );
  }
}
