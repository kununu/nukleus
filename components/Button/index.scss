@import '~kununu-theme/scss/base/variables';

// Common styles
.button {
  border-radius: $border-radius-sm;
  font-weight: $font-weight-semibold;
  box-shadow: none;
  text-transform: uppercase;
  display: inline-block;
  text-align: center;
  vertical-align: middle;
  touch-action: manipulation;
  cursor: pointer;
  border: 1px solid transparent;
  white-space: nowrap;
  padding: 10px 12px;
  transition: background .15s;
  font-size: $font-size-base;

  &,
  &:hover,
  &:focus,
  &.focus,
  &:active,
  &.active {
    text-decoration: none;
  }

  &:focus,
  &.focus {
    outline: 5px auto -webkit-focus-ring-color;
    outline-offset: -2px;
  }

  &.disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &.focus,
    &:active,
    &.active {
      cursor: not-allowed;
      opacity: .65;
    }
  }

  &.fullWidth {
    width: 100%;
  }
}

@media (max-width: $screen-sm-min) {
  .button {
    white-space: normal;

    .mobileFullWidth {
      width: 100%;
    }
  }
}

@mixin button-default(
  $color,
  $background,
  $hover-background : lighten($background, 5%),
  $active-background: darken($background, 5%)
) {
  color: $color;
  background-color: $background;

  &:hover,
  &:focus,
  &.focus {
    color: $color;
    background-color: $hover-background;
  }

  &:active,
  &.active {
    color: $color;
    background-color: $active-background;
  }

  &.disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &.focus,
    &:active,
    &.active {
      color: $color;
      background-color: $background;
    }
  }
}

@mixin button-outline(
  $color,
  $background,
  $hover-color: darken($color, 5%),
  $active-color: lighten($color, 5%)
) {

  &.outline {
    color: $color;
    background-color: $background;
    border: 1px solid $color;

    &:hover,
    &:focus,
    &.focus {
      color: $hover-color;
      border-color: $hover-color;
    }

    &:active,
    &.active {
      color: $active-color;
      border-color: $active-color;
    }

    &.disabled,
    &[disabled] {
      &,
      &:hover,
      &:focus,
      &.focus,
      &:active,
      &.active {
        color: $color;
        background-color: $background;
      }
    }
  }
}

// Apply the mixin to the buttons
.default {
  @include button-default($btn-default-color, $btn-default-bg);
  @include button-outline($btn-default-color, $btn-info-bg);
}

.primary {
  @include button-default($btn-primary-color, $btn-primary-bg);
  @include button-outline($btn-primary-bg, $btn-primary-color);
}

.secondary {
  @include button-default($btn-secondary-color, $btn-secondary-bg);
  @include button-outline($btn-secondary-bg, $btn-secondary-color);
}

.info {
  @include button-default($btn-info-color, $btn-info-bg);
  @include button-outline($btn-info-color, $btn-info-bg);
}

.danger {
  @include button-default($btn-danger-color, $btn-danger-bg);
  @include button-outline($btn-danger-bg, $btn-danger-color);
}

.link {
  @include button-default($btn-secondary-bg, $btn-info-bg);
  @include button-outline($btn-secondary-bg, $btn-info-bg);

  &:hover {
    text-decoration: underline;
    border-color: transparent;
  }

  &.disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &.focus,
    &:active,
    &.active {
      text-decoration: none;
    }
  }
}

.disabledLink {
  cursor: not-allowed;

  a {
    pointer-events: none;
  }
}
