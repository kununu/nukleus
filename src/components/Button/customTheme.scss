.button {
  color: white;
  border-radius: 5px;
  border: 0;
  text-transform: uppercase;
  font-weight: bold;
  padding: 10px 20px;
  box-shadow: 0px 5px 40px -11px rgba(0,0,0,0.76);
}

.primary {
  background-color: pink;
}