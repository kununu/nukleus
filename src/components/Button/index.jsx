import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import ThemeContext from 'utils/themeContext';
import themeable from 'utils/theming';

import styles from './index.scss';

export default class Button extends React.Component {
  static propTypes = {
    customTheme: PropTypes.string,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    htmlType: PropTypes.oneOf([
      'button',
      'submit',
      'reset',
    ]),
    link: PropTypes.element,
    mobileFullWidth: PropTypes.bool,
    onClick: PropTypes.func,
    outline: PropTypes.bool,
    text: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.string,
    ]),
    title: PropTypes.string,
    type: PropTypes.oneOf(['default', 'primary', 'secondary', 'info', 'danger', 'link', 'custom']),
    id: PropTypes.string,
  };

  static defaultProps = {
    customTheme: '',
    disabled: false,
    fullWidth: false,
    htmlType: 'submit',
    link: null,
    mobileFullWidth: false,
    onClick: null,
    outline: false,
    text: '',
    title: null,
    type: 'primary',
    id: null,
  };

  onClick = (e) => {
    const {
      disabled,
      onClick,
    } = this.props;

    e.preventDefault();

    if (disabled) return;

    onClick();
  }

  render () {
    const {
      customTheme,
      disabled,
      fullWidth,
      link,
      mobileFullWidth,
      onClick,
      outline,
      text,
      htmlType,
      title,
      type,
      id,
    } = this.props;

    const classes = classNames({
      button: true,
      fullWidth: fullWidth,
      mobileFullWidth: mobileFullWidth,
      outline: outline,
      [type === 'custom' ? customTheme : type]: true,
    }).split(' ')

    return (
      <ThemeContext.Consumer>
        {(context) => {
          const theme = themeable({...styles, ...context})

          const props = {
            className: theme(...classes),
            disabled,
            title,
            onClick: onClick && this.onClick,
            id,
          };

          if (link) {
            return (
              <div className={disabled ? styles.disabledLink : ''}>
                {React.cloneElement(link, props)}
              </div>
            );
          }

          return (
            <button // eslint-disable-line react/button-has-type
              {...props}
              type={htmlType}
            >
              {text}
            </button>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
