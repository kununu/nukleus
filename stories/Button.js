import React from 'react';
import {storiesOf} from '@storybook/react';

import Button from '../src/components/Button';

storiesOf('Test', module)
  .add('button', () => (
    <Button text="hi" />
  ));
