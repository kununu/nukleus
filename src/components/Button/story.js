import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {action} from '@storybook/addon-actions';
import {withInfo} from '@storybook/addon-info';
import {withKnobs, text, boolean} from '@storybook/addon-knobs/react';

import ThemeProvider from '../ThemeProvider';

import customTheme from './customTheme.scss';

import Button from './index';

storiesOf('Button', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('The button component!')(story)(context))
  .add('basics', () => (
    <>
      <h3>
        Primary button
      </h3>
      <Button
        type="primary"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />

      <hr />

      <h3>
      Secondary button
      </h3>
      <Button
        type="secondary"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />

      <hr />

      <h3>
      Info button
      </h3>
      <Button
        type="info"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />

      <hr />

      <h3>
      Danger button
      </h3>
      <Button
        type="danger"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />

      <hr />

      <h3>
      Default button
      </h3>
      <Button
        type="default"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
      />

      <hr />

      <h3>
      Button that is an link
      </h3>
      <Button
        type="primary"
        link={<a href="/">{text('link text', 'I am a link!')}</a>}
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />

      <hr />

      <h3>
      Custom button with no styling
      </h3>
      <Button
        type="custom"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />

      <hr />
    </>
  ))
  .add('custom', () => (
    <ThemeProvider theme={customTheme}>
      <h3>Custom theme for buttons</h3>

      <Button
        type="primary"
        disabled={boolean('disabled', false)}
        fullWidth={boolean('fullWidth', false)}
        mobileFullWidth={boolean('mobileFullWidth', false)}
        onClick={action('click')}
        outline={boolean('outline', false)}
        text={text('text', 'button')}
        title={text('title', 'This is some info about the button')}
      />
    </ThemeProvider>
  ));
