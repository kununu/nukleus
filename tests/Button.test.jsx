import React from 'react';
import {Link} from 'react-router';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import Button from 'Button'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

test('Renders button without crashing', () => {
  const component = renderer.create(
    <Button text="Test" onClick={() => {}} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders simple anchor button without crashing', () => {
  const component = renderer.create(
    <Button
      text="Test"
      link={<a href="/">Button</a>} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders react-router anchor button without crashing', () => {
  const component = renderer.create(
    <Button
      text="Test"
      link={<Link to={{pathname: '/test'}}>Button</Link>} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a disabled Button', () => {
  const component = renderer.create(
    <Button text="Test" onClick={() => {}} disabled />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders Button with outline', () => {
  const component = renderer.create(
    <Button text="Test" onClick={() => {}} outline />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a full width button', () => {
  const component = renderer.create(
    <Button text="Test" onClick={() => {}} fullWidth />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


test('Renders a mobile full width button', () => {
  const component = renderer.create(
    <Button text="Test" onClick={() => {}} mobileFullWidth />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a custom theme button', () => {
  const component = renderer.create(
    <Button text="Test" onClick={() => {}} type="custom" customTheme="customCssModuleClass" />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Fires onClick function when clicked', () => {
  const spyFunc = jest.fn();
  const component = mount(
    <Button text="Test" onClick={spyFunc} />
  );

  component.find('.button').simulate('click');
  expect(spyFunc).toHaveBeenCalled();
});

test('Does not fire onClick function when clicked if the button is disabled', () => {
  const spyFunc = jest.fn();
  const component = mount(
    <Button text="Test" onClick={spyFunc} disabled />
  );

  component.find('.button').simulate('click');
  expect(spyFunc).not.toHaveBeenCalled();
});
