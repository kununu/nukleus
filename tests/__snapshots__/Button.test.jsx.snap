exports[`test Renders Button with outline 1`] = `
<button
  className="button outline primary"
  disabled={undefined}
  onClick={[Function]}>
  Test
</button>
`;

exports[`test Renders a custom theme button 1`] = `
<button
  className="button customCssModuleClass"
  disabled={undefined}
  onClick={[Function]}>
  Test
</button>
`;

exports[`test Renders a disabled Button 1`] = `
<button
  className="button primary"
  disabled={true}
  onClick={[Function]}>
  Test
</button>
`;

exports[`test Renders a full width button 1`] = `
<button
  className="button fullWidth primary"
  disabled={undefined}
  onClick={[Function]}>
  Test
</button>
`;

exports[`test Renders a mobile full width button 1`] = `
<button
  className="button mobileFullWidth primary"
  disabled={undefined}
  onClick={[Function]}>
  Test
</button>
`;

exports[`test Renders anchor button without crashing 1`] = `
<a
  className="button primary"
  disabled={undefined}
  onClick={[Function]}
  style={Object {}}>
  Test
</a>
`;

exports[`test Renders button without crashing 1`] = `
<button
  className="button primary"
  disabled={undefined}
  onClick={[Function]}>
  Test
</button>
`;
